nspawn: support passing wireless interface to container

yuwata opened the pull request:
Closes #7873.

yuwata commented:
I have not tested this.

poettering commented:
looks superficially good to me.

yuwata commented:
This becomes much bigger than I initially expected. But, now renaming is also supported, and test cases are added. PTAL.

yuwata commented:
@Edu4rdSHL Please test again.

yuwata commented:
(BTW, this is nice article: https://www.suse.com/c/creating-virtual-wlan-interfaces/)

Edu4rdSHL commented:
I have tested it and everything works as expected. I did a few tests:

- Passing the interface just to get network access: OK
- Passing the interface to do WiFi pentesting tests (which include putting the interface in monitor mode, etc): OK
- Passing the interface to create custom network configs inside the container: OK

poettering commented on src/nspawn/nspawn-network.c:
```diff
@@ -544,6 +547,43 @@ int move_network_interfaces(int netns_fd, char **iface_pairs) {
+        if (strv_isempty(interface_pairs))
+                return 0;
+
+        r = namespace_open(getpid_cached(),
```
you can just pass 0 here as first arg

poettering commented on src/nspawn/nspawn-network.c:
```diff
@@ -541,6 +543,173 @@ static int needs_rename(sd_netlink **rtnl, sd_device *dev, const char *name) {
+
+        /* The command NL80211_CMD_SET_WIPHY_NETNS below passes phy instead of network interface,
+         * and does not take an interface name in the passed network namespace. Hence, we need to
+         * rename the interface before passing phy to the netns. */
```
so, if another interface already exists under the target name in the source namespace then the whole operation will fail, no? That would suck, no?

Maybe create a throw-away middle netns? i.e. open a new netns, migrate the iface there, then rename it in there, and then migrate it to the target ns? that way the interface will only exist under the old name in the source netns, and only under the new name in the target netns.

or am i missing something here, and this is not an issue?

yuwata replied:
Sounds nice. I will try.

yuwata commented:
@poettering Thank you for the review. Addressed all suggestions. PTAL.

@Edu4rdSHL Thank you for testing. If you'd like, please test again. Hopefully, this still works file.

poettering commented on src/nspawn/nspawn-network.c:
```diff
@@ -503,6 +506,58 @@ int test_network_interfaces_initialized(char **iface_pairs) {
+        return 0;
+}
+
+static int netns_fork(int netns_fd, int *ret_original_netns_fd) {
```
maybe call this fork_and_wait()? just to make clear that this is synchronous in the parent

poettering commented on src/nspawn/nspawn-network.c:
```diff
@@ -503,6 +506,58 @@ int test_network_interfaces_initialized(char **iface_pairs) {
         return 0;
 }
 
+static int netns_fork_child(int netns_fd, int *ret_original_netns_fd) {
```
maybe call this netns_child_begin() or so

the reason i am asking for this is because i was confused as first, because this function is after all not the full child, but just the initialization of the child, the main part of the child comes after this function returns after all. maybe a more precise name might help the next reader

poettering commented:
lgtm, just some superficial comments about function names.

yuwata commented:
@poettering Thank you for the review. The functions are renamed. Upgrading the green label.
